dask/dataframe/core.py

```diff
@@ -4555,14 +4556,80 @@ def sort_values(
             **kwargs,
         )
 
+    def _check_index(self, other: Any) -> Optional[Any]:
+        """Internal method to check the column passed to set_index.
+
+        Returns None if this column is already the index.
+        """
+
+        # Check other can be translated to column name or column object
+        index = other
+        if not isinstance(index, str):
+            if isinstance(index, Sequence):
+                # Accept ["a"], but not [["a"]]
+                if len(index) == 1 and (
+                    isinstance(index[0], str) or not isinstance(index[0], Sequence)
+                ):
+                    index = index[0]
+                else:
+                    raise NotImplementedError(
+                        "Dask dataframe does not yet support multi-indexes.\n"
+                        f"You tried to index with this index: {index}\n"
+                        "Indexes must be single columns only."
+                    )
+
+            # Can also be a frame
+            elif isinstance(index, DataFrame):
+                raise NotImplementedError(
+                    "Dask dataframe does not yet support multi-indexes.\n"
+                    f"You tried to index with several columns: {list(index.columns)}\n"
+                    "Indexes must be single columns only."
+                )
+
+        # Now index should be either a series...
+        if isinstance(index, Series):
+            # If it's the current index, nothing to do
+            if index._name == self.index._name:
+                return None
+            # Otherwise, should have same length than the index
+            if (len_index := len(index)) != (len_self := len(self)):
+                raise ValueError(
+                    f"Length mismatch: series passed has {len_index} rows, but data have {len_self} rows"
+                )
+            s = index
+
+        # ... or the name of a column/index
+        else:
+            # If already the index name, no-op
+            if index == self.index.name:
+                return None  # no-op, but note the difference: name checked instead of _name
+            # Should be a column name
+            try:
+                s = self[index]
+            except Exception:  # general exception cause not necessarily a KeyError (e.g. TypeError if datetime-indexed)
+                # Raise a KeyError regardless of the specific error that just happened
+                raise KeyError(
+                    f"Column {index} not found in data: use one of {list(self.columns)}"
+                )
+
+        # Ensure there are no nulls
+        if s.isna().any().compute():
```

Comment:
I realize that it's probably not possible to check for `na` without doing a `compute()`. However, I just want to flag that this will eagerly compute the index, which may or may not be otherwise desired.

Reply:
Yeah, I couldn't figure out any other way of doing this. Anyway, since it's an `any()` it should be light-weight, because as soon as there's a `True`, there's no need to calculate anything else. But I'm not really sure it stops there...

Reply:
@jsignell I'm not sure what happens here, and if this `compute()` will be a problem (I'm thinking of the recent `vindex` issue). Do you have any insight?

Reply:
Here it's not as with the `len`, where we know that if setting the index to an existing column, the `len` doesn't need to be checked. We must always check this, as there may be nulls in any column, or in the series provided directly by the user.

Anyway, AFAIK `set_index` always triggers computations (calculating quantiles, etc.), except for the very particular case where `sorted=True` and `divisions` are explicitly provided by the user. So I'd say it's reasonable that there's a `.compute()` here. Do you agree?